README.md
# Obsidian Vim Gutter Marker

This plugin adds gutter markers for *vim marks*.

![Vim Gutter Marker screenshot](./media/vim-gutter-marker.png)

## Settings

If line numbers are activated, the markers can be shown either before or after them.

## Known limitations

When switching to an "empty file" in the same pane and switching back, the gutter markers don't get reset.

If marks get deleted, this is not reflected in the gutter. The same thing applies to reusing the same mark character.

When opening a new file 

## Installation

The plugin can be installed manually or via BRAT. It is not in the community plugins.

## Credits

The gutter implementation is based on [Obsidian lapel](https://github.com/liamcain/obsidian-lapel/blob/dce7a1d9fc8ac9a2c8d3589b0e4f92d1f0241f39/src/headingWidget.ts).

The key capture is adapted from [this PR comment](https://github.com/mrjackphil/obsidian-jump-to-link/issues/35#issuecomment-1085905668).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.9](https://github.com/kometenstaub/obsidian-vim-gutter-marker/compare/0.0.8...0.0.9) (2022-07-15)


### Bug Fixes

* use correct API; improve settings ([b418092](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/b41809261cb28d4af8d4c20674ecbedad26a68d2))

### [0.0.8](https://github.com/kometenstaub/obsidian-vim-gutter-marker/compare/0.0.7...0.0.8) (2022-07-13)

### [0.0.7](https://github.com/kometenstaub/obsidian-vim-gutter-marker/compare/0.0.6...0.0.7) (2022-06-23)


### Bug Fixes

* use better API, error handling ([2161cb4](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/2161cb4f7e3b8464369e1d376521db8b708875ff))

### [0.0.6](https://github.com/kometenstaub/obsidian-vim-gutter-marker/compare/0.0.5...0.0.6) (2022-06-17)


### Bug Fixes

* ignore capital key if first ([d0077cc](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/d0077cc519028d780eb61ea29bdc5f39b334d844))

### [0.0.5](https://github.com/kometenstaub/obsidian-vim-gutter-marker/compare/0.0.4...0.0.5) (2022-06-17)


### Bug Fixes

* abort in reading mode ([ddb05c3](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/ddb05c392681c1e4aeb8ca112ef597b5b4689bd4))

### [0.0.4](https://github.com/kometenstaub/obsidian-vim-gutter-marker/compare/0.0.3...0.0.4) (2022-06-17)


### Bug Fixes

* Esc after m possible ([851e22d](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/851e22dbc482669e35b285f19c20c00f55f15b25))

### [0.0.3](https://github.com/kometenstaub/obsidian-vim-gutter-marker/compare/0.0.2...0.0.3) (2022-06-16)

### [0.0.2](https://github.com/kometenstaub/obsidian-vim-gutter-marker/compare/0.0.1...0.0.2) (2022-06-16)


### Bug Fixes

* clean up old event listeners ([0e154a8](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/0e154a8083297681867d1dd9c6010c04297f1f2f))
* marks diappear after opening other file in ([d2e8917](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/d2e8917114201af6e7be8cf935e7d570a3f4c613))

### 0.0.1 (2022-06-16)


### Features

* add CSS ([d71ce21](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/d71ce212743d98c39bc4147e3844c0f99f32e14a))
* cm6 yank highlighter ([c4cc296](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/c4cc296623ad512023b10525b5ed6a60622d575e))
* emit event for vim yank ([f9ced10](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/f9ced10f75f614ddc381f341652d9e912808bf8b))
* initial gutter implementation ([b7a115c](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/b7a115c495c217754cf7c6d6a4de16911355c5a9))
* restore gutter marks when marks present ([745480b](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/745480b2d94c1fb6cc432e46b2fd5b5d15276f39))
* setting, cleanup ([25db3bf](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/25db3bfba4db8684dd431c86782907bc1a7015eb))


### Bug Fixes

* CSS ([15c9a96](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/15c9a96bd7582db1fe98feb3d10c733979ee6c52))
* error handling, better pane handling ([69501fd](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/69501fd8fbc4ed9dc4beb1d1fedf4a2f954608ce))
* more typos ([25102e2](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/25102e29cb728038cb8dd6710c6040c844e53648))
* only add pane when not existant ([1b2f0db](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/1b2f0db6f47e5ef5c1440d14877114346a1a9cd6))
* only set marks in normal mode, don't remove ([78faa8b](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/78faa8bf7a30050ca3798cd28aa095b71dd6235e))
* persist markers ([a68b74f](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/a68b74f2d19c47712acd9859ae3557a461e26673))
* remove marks when file changed ([6bf5fcb](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/6bf5fcbc6e3b0ef17b71f83418bef923e810cd31))
* reset marks properly ([e924c8f](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/e924c8f9a64806af7548848fb3192df5ac6b0efe))
* show in pop-out windows; update docs ([915cedb](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/915cedb89d8f878ecdcb5af466ef4edcea63a870))
* trigger after removal ([c6f1090](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/c6f1090a7c2637fdc2df912663897ddaccc864ae))
* trigger typo ([949180f](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/949180f9feeded4afd7379ad5ef2bf5f948f074f))
* typo ([94a98f7](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/94a98f7831d47aef23f740b8814b4d04958d04a3))


### Performance

* use variable ([ab0a7c0](https://github.com/kometenstaub/obsidian-vim-gutter-marker/commit/ab0a7c00674d6aa7d99f05400cbd851cf0e6b2e4))


esbuild.mjs
// Thank you: https://github.com/aidenlx/media-extended/blob/main/esbuild.js
import esbuild from 'esbuild';
import fs from 'fs';
import process from 'process';
import builtins from 'builtin-modules';
import minify from 'css-minify';
import sass from 'sass';

const banner = `/*
THIS IS A GENERATED/BUNDLED FILE BY ESBUILD
If you want to view the source, visit the plugin’s github repository:
https://github.com/kometenstaub/obsidian-vim-gutter-marker
It is MIT-licensed:
	MIT License
	
	Copyright (c) 2022 kometenstaub and contributors
	
	Permission is hereby granted, free of charge, to any person obtaining a copy
	of this software and associated documentation files (the "Software"), to deal
	in the Software without restriction, including without limitation the rights
	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
	copies of the Software, and to permit persons to whom the Software is
	furnished to do so, subject to the following conditions:
	
	The above copyright notice and this permission notice shall be included in all
	copies or substantial portions of the Software.
	
	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
	SOFTWARE.
	
Other used code in this built file is licensed as follows:

The codemirror ViewPlugin is based on https://github.com/liamcain/obsidian-lapel/blob/dce7a1d9fc8ac9a2c8d3589b0e4f92d1f0241f39/src/headingWidget.ts which is MIT-licensed.
	MIT License

	Copyright (c) 2022 Liam Cain

	Permission is hereby granted, free of charge, to any person obtaining a copy
	of this software and associated documentation files (the "Software"), to deal
	in the Software without restriction, including without limitation the rights
	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
	copies of the Software, and to permit persons to whom the Software is
	furnished to do so, subject to the following conditions:

	The above copyright notice and this permission notice shall be included in all
	copies or substantial portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
	SOFTWARE.

*/
`;

const cssBanner = `/*
This plugin's CSS is MIT-licensed:
	Copyright 2022 kometenstaub
	Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
	documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
	rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
	persons to whom the Software is furnished to do so, subject to the following conditions:
	The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
	Software.
	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
	WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
	COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
	OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
	
*/
`;

const copyManifest = {
	name: 'copy-manifest',
	setup: (build) => {
		build.onEnd(() => {
			fs.copyFileSync('manifest.json', 'build/manifest.json');
		});
	},
};

const copyMinifiedCSS = {
	name: 'minify-css',
	setup: (build) => {
		build.onEnd(async () => {
			const { css } = sass.compile('src/styles.scss');
			const minCss = await minify(css);
			const content = `${cssBanner}\n${minCss}`;
			fs.writeFileSync('build/styles.css', content, {
				encoding: 'utf-8',
			});
		});
	},
};

const isProd = process.env.BUILD === 'production';

(async () => {
	try {
		await esbuild.build({
			entryPoints: ['src/main.ts', 'src/styles.scss'],
			bundle: true,
			watch: !isProd,
			external: [
				'obsidian',
				'electron',
				'@codemirror/autocomplete',
				'@codemirror/collab',
				'@codemirror/commands',
				'@codemirror/fold',
				'@codemirror/language',
				'@codemirror/lint',
				'@codemirror/search',
				'@codemirror/state',
				'@lezer/highlight',
				'@codemirror/view',
				...builtins,
			],
			format: 'cjs',
			target: 'es2018',
			banner: { js: banner },
			sourcemap: isProd ? false : 'inline',
			minify: isProd,
			logLevel: 'info',
			treeShaking: true,
			define: {
				'process.env.NODE_ENV': JSON.stringify(process.env.BUILD),
			},
			outdir: 'build/',
			plugins: [copyManifest, copyMinifiedCSS],
			loader: { '.scss': 'text' },
		});
	} catch (err) {
		console.error(err);
		process.exit(1);
	}
})();


src/main.ts
import {
	App,
	EventRef,
	Events,
	MarkdownView,
	Plugin,
	PluginSettingTab,
	Setting,
	View,
	WorkspaceLeaf,
} from 'obsidian';
import { EditorView, ViewPlugin, gutter, GutterMarker } from '@codemirror/view';
import { Prec, RangeSet, RangeSetBuilder } from '@codemirror/state';

// add type safety for the undocumented methods
declare module 'obsidian' {
	interface Vault {
		setConfig: (config: string, newValue: boolean) => void;
		getConfig: (config: string) => boolean;
	}
}

interface markData {
	mark: string;
	from: number;
	to: number;
}

interface VimMarkSettings {
	showBeforeLineNumbers: boolean;
}

const DEFAULT_SETTINGS: VimMarkSettings = { showBeforeLineNumbers: true };

class VimEvent extends Events {
	on(name: 'vim-setmark', callback: (data: markData[]) => void): EventRef;
	//on(name: 'vim-delmark', callback: (text: string) => void): EventRef;
	on(name: string, callback: (...data: any) => any, ctx?: any): EventRef {
		return super.on(name, callback, ctx);
	}
}

// based on https://github.com/liamcain/obsidian-lapel/blob/dce7a1d9fc8ac9a2c8d3589b0e4f92d1f0241f39/src/headingWidget.ts (MIT-licensed)
class MarkMarker extends GutterMarker {
	constructor(readonly view: EditorView, readonly marker: string) {
		super();
	}

	toDOM(view: EditorView): Node {
		const markEl = createFragment((frag) => {
			frag.createDiv({ text: this.marker });
		});
		return markEl;
	}

	elementClass = 'vim-gutter-mark';
}

// cm6 view plugin
// based on https://github.com/liamcain/obsidian-lapel/blob/dce7a1d9fc8ac9a2c8d3589b0e4f92d1f0241f39/src/headingWidget.ts (MIT-licensed)
function vimGutterMarker(app: App, evt: VimEvent, showBeforeLineNumbers: boolean) {
	const markers = ViewPlugin.fromClass(
		class {
			markers: RangeSet<MarkMarker>;

			constructor(public view: EditorView) {
				evt.on('vim-setmark', (data) => {
					this.markers = this.makeGutterMarker(view, data);
				});
			}
			//update unnecessary because highlight gets removed by timeout; otherwise it would never apply the classes
			//update(update: ViewUpdate) {
			//	if (!update.state.field(editorLivePreviewField)) {
			//		this.markers = RangeSet.empty;
			//		return;
			//	}
			//	this.markers = this.makeGutterMarker(this.view, this.oldData);
			//}

			makeGutterMarker(view: EditorView, data: markData[]) {
				const builder = new RangeSetBuilder<MarkMarker>();
				if (data.length === 0) {
					this.markers = RangeSet.empty;
					return builder.finish();
				}
				for (const el of data) {
					const dec = new MarkMarker(view, el.mark);
					builder.add(el.from, el.to, dec);
				}
				return builder.finish();
			}
		}
	);
	const gutterPrec = showBeforeLineNumbers ? Prec.high : Prec.low;
	return [
		markers,
		gutterPrec(
			gutter({
				class: 'cm-vim-mark',
				markers(view) {
					return view.plugin(markers)?.markers || RangeSet.empty;
				},
			})
		),
	];
}

export default class MarkGutter extends Plugin {
	settings: VimMarkSettings;
	marks: markData[] | [] = [];
	contentEl: HTMLElement;
	grabKey: (evt: KeyboardEvent) => void;
	oldLeaf: MarkdownView;
	leaves: Set<{ path: string; id: string; marks?: markData[] }> = new Set();
	oldContentEl: HTMLElement;

	async onload() {
		console.log('Vim Gutter Marker plugin loading.');
		await this.loadSettings();
		this.addSettingTab(new VimGutterSettingTab(this.app, this));
		if (this.app.vault.getConfig('vimMode')) {
			const vimEvent = new VimEvent();

			this.registerEditorExtension(
				vimGutterMarker(app, vimEvent, this.settings.showBeforeLineNumbers)
			);

			this.registerEvent(
				app.workspace.on('file-open', async (file) => {
					if (this.oldContentEl) {
						this.oldContentEl.removeEventListener('keydown', this.grabKey, {
							capture: true,
						});
					}
					// reset marks for new pane; get them from history later, if available
					this.marks = [];
					const currentLeaf = app.workspace.getActiveViewOfType(MarkdownView);
					if (!currentLeaf) {
						return;
					}
					//@ts-expect-error,...
					const currentId: string = app.workspace.getActiveViewOfType(View).leaf.id;
					if (!currentId) return;
					const leaves = Array.from(this.leaves);
					// focus changed between panes, but file in pane didn't change, so old marks are still there,
					// but not in the gutter anymore - add old marks back
					if (this.oldLeaf !== currentLeaf) {
						const result = leaves.find((el) => {
							if (el.id === currentId) {
								if (el.marks) {
									this.marks = el.marks;
									vimEvent.trigger('vim-setmark', this.marks);
								}
								return true;
							}
						});
						// new leaf, old ones don't get added
						if (!result) {
							this.leaves.add({
								path: file.path,
								id: currentId,
							});
						}
					}
					// check if there are still marks in the same leaf
					// this can be the case when only the file changed, but no other
					// leaf was opened
					if (this.oldLeaf === currentLeaf) {
						//if (currentLeaf.getViewType() ===)
						const myMarks = await currentLeaf.editor.cm.cm.marks;
						const markLength = Object.keys(myMarks).length;
						if (markLength === 0) {
							// otherwise cm6 remembers them for the editor in which marks
							// were set before
							this.marks = [];
							vimEvent.trigger('vim-setmark', this.marks);
						}
					}
					this.oldLeaf = app.workspace.getActiveViewOfType(MarkdownView);

					this.contentEl = currentLeaf.contentEl;
					if (!this.contentEl) {
						return;
					}
					// for removing the event listener on the next file-open event
					this.oldContentEl = currentLeaf.contentEl;
					// adapted from: https://github.com/mrjackphil/obsidian-jump-to-link/issues/35#issuecomment-1085905668
					let keyArray: string[] = [];
					this.grabKey = (event: KeyboardEvent) => {
						if (currentLeaf.getMode() === 'preview') {
							return;
						}

						// empty array if Esc
						if (event.key === 'Escape') {
							keyArray = [];
							return;
						}

						// test if keypress is capitalized
						if (/^[a-z]$/i.test(event.key)) {
							const isCapital = event.shiftKey;
							if (!isCapital && event.key !== 'm' && keyArray.length === 0) {
								return;
							}
							if (isCapital && keyArray.length === 0) {
								return;
							}
							if (isCapital && keyArray.length === 1) {
								// capture uppercase
								keyArray.push(event.key.toUpperCase());
							} else {
								// capture lowercase
								keyArray.push(event.key);
							}
						}

						// stop when length of array is equal to 2
						if (keyArray.length === 2) {
							const mode =
								// @ts-expect-error, not typed
								window.CodeMirrorAdapter.Vim.maybeInitVimState_(
									// @ts-expect-error, not typed
									app.workspace.getActiveViewOfType(MarkdownView).editor.cm.cm
								).mode;
							// the mode is not always set, even when it is active
							if (mode === undefined || mode === 'normal') {
								const { editor } = app.workspace.getActiveViewOfType(MarkdownView);
								const intCursorFrom = editor.getCursor('from');
								intCursorFrom.ch = 0;
								const intCursorTo = editor.getCursor('to');
								intCursorTo.ch = 0;
								const cursorFrom = editor.posToOffset(intCursorFrom);
								const cursorTo = editor.posToOffset(intCursorTo);

								this.marks.push({
									mark: keyArray.at(1),
									from: cursorFrom,
									to: cursorTo,
								});
								this.marks.sort(
									(
										a: {
											mark: string;
											from: number;
											to: number;
										},
										b: {
											mark: string;
											from: number;
											to: number;
										}
									) => {
										if (a.from < b.from) {
											return -1;
										}
										if (a.from > b.from) {
											return 1;
										}
										return 0;
									}
								);
								// for later comparison
								const leaves = Array.from(this.leaves);
								const currentEl = leaves.find((el) => {
									if (el.id === currentId) {
										return true;
									}
								});

								currentEl['marks'] = this.marks;
								leaves.push(currentEl);
								this.leaves = new Set(leaves);
								vimEvent.trigger('vim-setmark', this.marks);
							}
							keyArray = [];
							// removing eventListener after proceeded
							//contentEl.removeEventListener("keydown", grabKey, { capture: false })
						}
					};
					this.contentEl.addEventListener('keydown', this.grabKey, {
						capture: true,
						passive: true,
					});
				})
			);
		}
	}
	async onunload() {
		if (this.oldContentEl) {
			this.oldContentEl.removeEventListener('keydown', this.grabKey, {
				capture: true,
			});
		}
		console.log('Vim Gutter Marker plugin unloaded.');
	}
	async loadSettings() {
		this.settings = Object.assign({}, DEFAULT_SETTINGS, await this.loadData());
	}

	async saveSettings() {
		await this.saveData(this.settings);
	}
}

class VimGutterSettingTab extends PluginSettingTab {
	plugin: MarkGutter;

	constructor(app: App, plugin: MarkGutter) {
		super(app, plugin);
		this.plugin = plugin;
	}

	display(): void {
		const { containerEl } = this;
		const { settings } = this.plugin;

		containerEl.empty();

		containerEl.createEl('h2', { text: 'Vim Gutter Marker settings' });

		new Setting(containerEl)
			.setName('Show marker before line numbers')
			.setDesc(
				'If enabled, the markers are to the left of the line numbers. If disabled, they are to the right. Requires a reload to take effect.'
			)
			.addToggle((toggle) => {
				toggle.setValue(settings.showBeforeLineNumbers).onChange(async (state) => {
					settings.showBeforeLineNumbers = state;
					await this.plugin.saveSettings();
				});
			});
	}
}
